packages/fxa-auth-server/lib/payments/stripe-firestore.ts

```diff
@@ -0,0 +1,211 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { CollectionReference, Firestore } from '@google-cloud/firestore';
+import { Stripe } from 'stripe';
+
+export enum FirestoreStripeError {
+  FIRESTORE_CUSTOMER_NOT_FOUND = 'FirestoreCustomerNotFound',
+  FIRESTORE_SUBSCRIPTION_NOT_FOUND = 'FirestoreSubscriptionNotFound',
+  FIRESTORE_INVOICE_NOT_FOUND = 'FirestoreInvoiceNotFound',
+  STRIPE_CUSTOMER_MISSING_UID = 'StripeCustomerMissingUid',
+  STRIPE_CUSTOMER_DELETED = 'StripeCustomerDeleted',
+}
+
+function newFirestoreStripeError(
+  message: string,
+  code: FirestoreStripeError
+): Error {
+  const error = new Error(message);
+  error.name = code;
+  return error;
+}
+
+export class StripeFirestore {
+  constructor(
+    private firestore: Firestore,
+    private customerCollectionDbRef: CollectionReference,
+    private stripe: Stripe
+  ) {}
+
+  /**
+   * Get a Stripe customer by id, and insert it into Firestore keyed to the fxa uid.
+   *
+   * This method is used for populating the customer if missing from Stripe and also
+   * loads the customers subscriptions into Firestore.
+   */
+  async fetchAndInsertCustomer(customerId: string) {
+    const customer = await this.stripe.customers.retrieve(customerId, {
+      expand: ['subscriptions'],
+    });
+    if (customer.deleted) {
+      throw newFirestoreStripeError(
+        `Customer ${customerId} was deleted`,
+        FirestoreStripeError.STRIPE_CUSTOMER_DELETED
+      );
+    }
+    // Drop Stripe.Response attribute from API call.
+    delete (customer as any).lastResponse;
+
+    const uid = customer.metadata.userid;
+    if (!uid) {
+      throw newFirestoreStripeError(
+        `Customer ${customerId} has no uid`,
+        FirestoreStripeError.STRIPE_CUSTOMER_MISSING_UID
+      );
+    }
+
+    const inserts = [this.insertCustomerRecord(uid, customer)];
+    if (customer.subscriptions) {
+      for (const subscription of customer.subscriptions.data) {
+        inserts.push(
+          this.customerCollectionDbRef
+            .doc(uid)
+            .collection('subscriptions')
+            .doc(subscription.id)
+            .set(subscription, { merge: true })
+        );
+      }
+    }
+    await Promise.all(inserts);
+    return customer;
+  }
+
+  /**
+   * Insert a Stripe customer into Firestore keyed to the fxa id.
+   */
+  insertCustomerRecord(uid: string, customer: Stripe.Customer) {
+    return this.customerCollectionDbRef.doc(uid).set(customer, { merge: true });
+  }
+
+  /**
+   * Insert a subscription record into Firestore under the customer's stripe id.
+   */
+  async insertSubscriptionRecord(subscription: Stripe.Subscription) {
```

Comment:
This function does not appear to be used anywhere (aside from tests).

Reply:
Correct, that's true for all of these. They might change slightly in the next PR where they're used.